### DISCUSSION TITLE
Widget Font problems

### DISCUSSION THREAD

**Discussion by nkmhor** [pymupdf/PyMuPDF#1697] [Looking for help]
We are trying to create a PDF form with form (widget) fields that get filled in using PyMuPDF.  The PDF form is created in Adobe Acrobat.  The python code sets each widget field_value, adds read only to its field_flags, and prepends a fixed string to each field_name.  The form designer is picking fonts in Acrobat from the list of 14 base fonts supported in PyMuPDF.  PyMuPDF keeps complaining about fonts not being found.  For example, if Courier and Courier-Bold are used in the form fields, the Courier fields show up as Courier but the Courier Bold fields show up as Helvetica in Acrobat and in printed output.  The Helvetica fields 'turn into' Courier Bold when they are clicked on in Acrobat but revert to Helvetica when focus moves off of them.  We want to be able to see Courier Bold in the printed output of the form after the form is completed by the python code.  Ideally, we also want to be able to use more than the 14 base fonts.  Can anybody explain what is happening and how the form designer can specify fonts for form fields and/or what the software developer can do to preserve the fonts selected by the form designer?

**Comment by JorjMcKie** [comment]
The reason should be that PyMuPDF only supports the basic weights of the Base-14 fonts - no bold, no italic.

**Reply by nkmhor** [reply]
Hello Jorj, Thanks for the reply.  The documentation says PyMuPDF supports the fonts below which includes some bold and italic fonts.  The documentation also says, "We currently support the 14 PDF base fonts using the following fixed reference names, or any name of an **already existing field font**. When specifying a text font for new or changed widgets, either choose one in the first table column (upper and lower case supported), or one of the already existing form fonts. In the latter case, spelling must exactly match."  The phrase "**or any name of an already existing field font**" is of interest to me.  How does the form designed specify the font in Adobe Acrobat (under Microsoft Windows) so that it will render properly after python fills in the form fields using PyMuPDF?  For example, "Courier New" is a Microsoft Windows font; "Cour" and "Courier" are not.  When "Cour" is specified as a font in Adobe Acrobat, I get the following message from PyMuPDF, "Unable to locate PDF font substituting Helvetica" for font name=Cour.


```
Reference | Base14 Fontname
CoBI      | Courier-BoldOblique
CoBo      | Courier-Bold
CoIt      | Courier-Oblique
Cour      | Courier
HeBI      | Helvetica-BoldOblique
HeBo      | Helvetica-Bold
HeIt      | Helvetica-Oblique
Helv      | Helvetica (default)
Symb      | Symbol
TiBI      | Times-BoldItalic
TiBo      | Times-Bold
TiIt      | Times-Italic
TiRo      | Times-Roman
ZaDb      | ZapfDingbats

```

**Comment by JorjMcKie** [comment]
There is a documentation issue: I forgot to update some parts to make clear, that only Times-Roman, Helvetica and Cour are supported for adding or updating fields - no bold, no italic.
MuPDF dropped support for the other weights in v1.16.

Otherwise please send me a reproducer for your problem. I think I do not understand it.

**Reply by nkmhor** [reply]
Hello Jorj,  If only Times-Roman, Helvetica and Cour are supported for adding or updating fields - no bold, no italic then I guess I have my answer for PyMuPDF.  Are you aware of any other [free] packages that support PDF fields or specific documentation on how to set PDF form fields using other fonts if I wanted to do it from python outside of PyMuPDF?
